In [1]:
import sys, os
sys.path.append(os.pardir) # Why we do this? We want to import the parent's folder
from dataset.mnist import load_mnist # import dataset

# First call this dataset, it will cost long time ...
(x_train, t_train), (x_test, t_test) = load_mnist(flatten=True, normalize=False)

# print the data shape
print(x_train.shape)

(60000, 784)


In [3]:
# print the data shape
print(t_train.shape)
print(x_test.shape)
print(t_test.shape)

(60000,)
(10000, 784)
(10000,)


#### load_mnist() function:
* It will return (x_train, t_train) (x_test, t_test). 
    * x: image
    * t: test
* load_mnist(normalize=True, flatten=True, one_hot_label=False)
    * normalize
        * `= True`: 0.0 ~ 1.0
        * `= False`: 0 ~ 255
        
    * flatten
        * `= True`: 1-dim array with 784 pixels
        * `= False`: 3-dim arrays with 1x28x28
        
    * one_hot_label
        * `= True`: The label is one-hot.
        * `= False`: normal

In [2]:
import sys, os
sys.path.append(os.pardir) # Why we do this? We want to import the parent's folder
from dataset.mnist import load_mnist # import dataset
import numpy as np
from PIL import Image # Python Image Library

def img_show(img):
    pil_img = Image.fromarray(np.uint8(img))
    pil_img.show()
    
(x_train, t_train), (x_test, t_test) = load_mnist(flatten=True, normalize=False)
img = x_train[0]
label = t_train[0]
print(label)

5


In [3]:
print(img.shape)
img = img.reshape(28, 28)  # change the image.shape to the original size
print(img.shape)

(784,)
(28, 28)


In [4]:
img_show(img)

# Neural Network for MNIST

* Input-Layer:   
    784 Nodes(pixel) = 28*28, because the image from the dataset of MNIST is always 28*28.

* Output-Layer:   
    10, because there are 10 numbers for decimalism.

* Hidden-Layer:  
    2 layers. The number of the nodes from these Hidden Layer is not important. We set here for 1st Hidden-Layer to 50, 2nd Hidden-Layer to 100. We can also set other number.

In [11]:
import pickle

def softmax(x):
    c = np.max(x)
    return np.exp(x-c) / np.sum(np.exp(x-c))

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def get_data():
    (x_train, t_train), (x_test, t_test) = load_mnist(normalize=True, flatten=True, one_hot_label=False)
    return x_test, t_test


def init_network():
    with open("sample_weight.pkl", 'rb') as f:  
        # the weights from learning will be saved in sample_weight.pkl
        # In this .pkl file, the weights and the bias are saved by dict{}.
        network = pickle.load(f)
    return network


def predict(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']
    
    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = softmax(a3)
    
    return y

In [12]:
x, t = get_data()
network = init_network()
accuracy_cnt = 0
for i in range(len(x)):
    y = predict(network, x[i])
    p = np.argmax(y) # get the index of the parameter, which has the maximal probability
    if p == t[i]:
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352


### It means 95.32% data from this dataset are correctly classified.